### ISSUE TITLE
Query performance degraded on x.39.2

### ISSUE THREAD

**Issue by szsuyuji** [metabase/metabase#16380]
**Describe the bug**
v0.39.3 and v0.39.2 has the same problem 

cost 21.5s
![image](https://user-images.githubusercontent.com/1931764/120992501-3e1c9280-c7b5-11eb-81dc-1c94e0c3321c.png)

presto cost 28ms
![image](https://user-images.githubusercontent.com/1931764/120997136-8b026800-c7b9-11eb-8c9a-73f1d946796b.png)

v0.39.1 cost 3.4s

![image](https://user-images.githubusercontent.com/1931764/120992627-5f7d7e80-c7b5-11eb-9d98-e42c802de5c1.png)

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by flamber** [comment]
On my TrinoDB, I'm seeing average query time of a simple table about 0.7s on 0.39.1 and 3.4s on 0.39.2 for 10 runs.

But I'm also noticing other database types having slower response. Like BigQuery on a simple table, is 1.7s vs 2.7s for 10 runs.
I have no idea what could have caused this, since nothing in the [commits sticks out](https://github.com/metabase/metabase/compare/v0.39.1...v0.39.2).

**Comment by jackchongs** [comment]
Facing the same problem,  v0.39.3 presto engine download data very slow

**Comment by bvallelunga** [comment]
We are also seeing very slow queries as well, where most dashboards are timing out. We are using Postgres.

**Comment by flamber** [comment]
@bvallelunga Please post "Diagnostic Info" from Admin > Troubleshooting, and how you're hosting Metabase (Docker, Kubernetes, etc).

**Comment by bvallelunga** [comment]
Here is the diagnostic info

``` json
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.11+9",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.11",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.11+9",
    "os.name": "Linux",
    "os.version": "4.19.167+",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.11"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.18"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-05-27",
      "tag": "v0.39.3",
      "branch": "release-x.39.x",
      "hash": "598a112"
    },
    "settings": {
      "report-timezone": "US/Pacific"
    }
  }
}
```

**Comment by flamber** [comment]
@bvallelunga Please describe how you are running Metabase - JAR, Docker, Kubernetes, ....?
The more details you provide, the better a chance there is to figuring out where to look and get the issue fixed.

**Comment by archer2012** [comment]
@flamber  I have the same question with presto engine . I use run metabase in centos with jar.  I always use presto to create view and metabase query this view . I don't kown Does it have anything to do with this

**Comment by yurkazaytsev** [comment]
@archer2012 I don't know if it helps, but it work in my case. I switched from java 1.8 to openJDK11 and it improves the fetching speed from presto.

**Comment by szsuyuji** [comment]
> @archer2012 I don't know if it helps, but it works in my case. I've switched from java 1.8 to openJDK11 and it improves the fetching speed from presto.

![image](https://user-images.githubusercontent.com/1931764/123884498-f2a57080-d97d-11eb-8cc5-f1ae8d83bdf2.png)
docker image is jdk11

**Comment by archer2012** [comment]
> @archer2012 I don't know if it helps, but it works in my case. I've switched from java 1.8 to openJDK11 and it improves the fetching speed from presto.

I've used jdk11, As you say it's improves the fetching speed from presto.  but Still slowlyer  than presto Command line query  when fetch limit 1000. it's seem metabase  Spend more than 3s to deal with 1000 data ?   And If the table's field the more the show data slowly more!

![image](https://user-images.githubusercontent.com/3227239/123887274-cb51a200-d983-11eb-804a-81b555eeae70.png)

![image](https://user-images.githubusercontent.com/3227239/123887546-4b780780-d984-11eb-99b6-c4a1aeb1eaa9.png)

![image](https://user-images.githubusercontent.com/3227239/123887665-824e1d80-d984-11eb-8eeb-e068c9145071.png)

**Comment by archer2012** [comment]
> > @archer2012 I don't know if it helps, but it works in my case. I've switched from java 1.8 to openJDK11 and it improves the fetching speed from presto.
> 
> I've used jdk11, As you say it's improves the fetching speed from presto. but Still slowlyer than presto Command line query when fetch limit 1000. it's seem metabase Spend more than 3s to deal with 1000 data ? And If the table's field the more the show data slowly more!
> 
> ![image](https://user-images.githubusercontent.com/3227239/123887274-cb51a200-d983-11eb-804a-81b555eeae70.png)
> 
> ![image](https://user-images.githubusercontent.com/3227239/123888158-744ccc80-d985-11eb-89e6-8a5bb2a092e4.png)
> 
> ![image](https://user-images.githubusercontent.com/3227239/123887665-824e1d80-d984-11eb-8eeb-e068c9145071.png)

@szsuyuji @flamber @yurkazaytsev 
I think there are some problems about metabase presto dirver , I use other data source like postgresql or mysql not have this problem. Can you reappear an give me some help ?  Our company depend on presto engine！

`{
  "browser-info": {
    "language": "zh-CN",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "Java(TM) SE Runtime Environment",
    "java.runtime.version": "11.0.11+9-LTS-194",
    "java.vendor": "Oracle Corporation",
    "java.vendor.url": "https://openjdk.java.net/",
    "java.version": "11.0.11",
    "java.vm.name": "Java HotSpot(TM) 64-Bit Server VM",
    "java.vm.version": "11.0.11+9-LTS-194",
    "os.name": "Linux",
    "os.version": "3.10.0-1062.12.1.el7.x86_64",
    "user.language": "en",
    "user.timezone": "Asia/Shanghai"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "presto"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "1.4.197 (2018-03-18)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "1.4.197 (2018-03-18)"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-06-15",
      "tag": "v0.39.4",
      "branch": "release-x.39.x",
      "hash": "f538050"
    },
    "settings": {
      "report-timezone": null
    }
  }
}`

**Comment by flamber** [comment]
@archer2012 
Revert/downgrade to 0.39.1.
And migrate away from H2: https://www.metabase.com/docs/latest/operations-guide/migrating-from-h2.html
There's work being done to change the Presto driver to use JDBC instead.

**Comment by archer2012** [comment]
> And migrate away from H2: https://www.metabase.com/docs/latest/operations-guide/migrating-from-h2.html

I know that , I just show you demo to explain this problem.  it's not work whatever use H2 or Mysql to  improves the fetching speed from presto.

**Comment by archer2012** [comment]
> @archer2012
> Revert/downgrade to 0.39.1.
> And migrate away from H2: https://www.metabase.com/docs/latest/operations-guide/migrating-from-h2.html
> There's work being done to change the Presto driver to use JDBC instead.

ok,it's work, thinks ，does v0.40 will fix this problem?

**Comment by flamber** [comment]
@archer2012 When this issue is closed, then a milestone will be attached. I don't know if we will finish the driver before 0.40

**Comment by szsuyuji** [comment]
> @archer2012 When this issue is closed, then a milestone will be attached. I don't know if we will finish the driver before 0.40

@archer2012 i try to use the presto driver of 0.39.1 ,it not work
![image](https://user-images.githubusercontent.com/1931764/125030529-d7dda500-e0bd-11eb-9923-9fc037304386.png)

**Comment by archer2012** [comment]
> > @archer2012 When this issue is closed, then a milestone will be attached. I don't know if we will finish the driver before 0.40
> 
> @archer2012 i try to use the presto driver of 0.39.1 , still slowly
> ![image](https://user-images.githubusercontent.com/1931764/125030529-d7dda500-e0bd-11eb-9923-9fc037304386.png)

en ......  I try this way too , still slowly . it's seem look like metabase core's problem.  I hope the author can solve this problem as soon as possible

**Comment by jeff303** [comment]
> On my TrinoDB, I'm seeing average query time of a simple table about 0.7s on 0.39.1 and 3.4s on 0.39.2 for 10 runs.
> 
> But I'm also noticing other database types having slower response. Like BigQuery on a simple table, is 1.7s vs 2.7s for 10 runs.
> I have no idea what could have caused this, since nothing in the [commits sticks out](https://github.com/metabase/metabase/compare/v0.39.1...v0.39.2).

Do you mean Trino or Presto?  I'm trying `sample-dataset` on Presto (basically the query results showing all `orders`), on `0.39.1`, running, killing the app, switching to `0.39.2`, relaunching the web server, running the same exact question, and finding the runtime steady (at around 3.4 sec in both cases).  If this is something you can easily repro, then please ping me so we can do a screen share.

**Comment by jeff303** [comment]
Profiling revealed that the calls to `Log4jLoggerFactory.getLogger` became much slower between the two versions.  See attached screenshots.

<img width="1234" alt="Presto-parse_row_fn-trace-v0 39 3" src="https://user-images.githubusercontent.com/3521562/125381107-f27f7a80-e358-11eb-9763-d6c0a5c941d5.png">
<img width="1234" alt="Presto-parse_row_fn-trace-v0 39 1" src="https://user-images.githubusercontent.com/3521562/125381111-f3181100-e358-11eb-926c-829160fb2d6d.png">

On a whim, I searched through any changes between 0.39.1 and 0.39.2 that could be relevant, and found only #16005 (based on the comment added there).  Sure enough, backing that change out seems to fix the problem; see #17002.

Unclear why this change (which was ostensibly to fix a performance warning) significantly worsened performance, but in any case, we will probably want to do this reversion and then separately investigate how to do a different fix for #13625.

**Comment by jeff303** [comment]
Should be resolved in the next 39 and 40 point releases.
